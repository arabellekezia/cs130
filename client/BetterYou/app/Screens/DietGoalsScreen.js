import React from "react";
import { SafeAreaView, StyleSheet, View, Text, Linking } from "react-native";
import { ScrollView } from "react-native-gesture-handler";
import AppText from "../components/AppText";
import AppTextInput from "../components/AppTextInput";
import TextButton from "../components/TextButton";
import TitleText from "../components/TitleText";
import GoalsService from "../services/GoalsService";


const MIN_CALORIES = 1000;

function DietGoalsScreen({ navigation }) {
  const [calorieGoal, setCalorieGoal] = React.useState(0);
  const [err, setError] = React.useState({ calorieBudget: false });

  async function save() {
    const validCalorieGoal =
      calorieGoal > MIN_CALORIES && Boolean(Number(calorieGoal));
    if (!validCalorieGoal) {
      setError({ calorieBudget: true });
      return;
    }
    await GoalsService.setCalorieGoal(calorieGoal);
    navigation.popToTop();
  }

  function resetErrors() {
    setError({ calorieBudget: false });
  }

  function displayErrorMessage(err) {
    if (err.calorieBudget) {
      if (calorieGoal <= MIN_CALORIES) {
        return (
          <ErrorMessage message="Your daily calorie budget must be greater than 1000 calories." />
        );
      }
      if (!Boolean(Number(calorieGoal))) {
        return (
          <ErrorMessage message="Your daily calorie budget must be a number." />
        );
      }
    }
  }

  return (
    <ScrollView
      contentContainerStyle={{
        flex: 1,
      }}
      keyboardShouldPersistTaps="handled"
    >
      <SafeAreaView style={styles.container}>
        <TitleText style={styles.pageTitle} children="Diet Goals" />
        <View style={styles.form}>
          <AppText
            style={styles.question}
            children="What is your daily calorie budget?"
          />
          <AppTextInput
            style={styles.textInput}
            placeholder="2000 cal"
            isError={err.calorieBudget}
            keyboardType="numeric"
            onChangeText={(calories) => {
              setCalorieGoal(calories);
              resetErrors();
            }}
          />
          {displayErrorMessage(err)}
          <HealthInfo style={styles.healthInfoContainer} />
        </View>
        <TextButton
          style={styles.saveButton}
          name="Save"
          onPress={async () => {
            await save();
          }}
        />
      </SafeAreaView>
    </ScrollView>
  );
}

function ErrorMessage({ message }) {
  return (
    <AppText
      style={{
        alignSelf: "flex-start",
        marginLeft: "2.5%",
        color: "red",
      }}
      testID="error-message"
      children={message}
    />
  );
}

function HealthInfo({ style }) {
  const calorieCalculatorLink =
    "https://www.calculator.net/calorie-calculator.html";
  return (
    <View style={style}>
      <Text style={styles.informationalTextHeader}>
        New to calorie counting?
      </Text>
      <Text style={styles.informationalText}>
        <Text style={styles.informationalText}>{"Use the following "}</Text>
        <Text
          style={styles.learnMoreText}
          onPress={() => {
            Linking.openURL(calorieCalculatorLink);
          }}
        >
          {"calorie calculator"}
        </Text>
        <Text>
          {
            " as a guideline to determine your daily calorie intake based on your weight loss goals!"
          }
        </Text>
      </Text>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: "white",
    alignItems: "center",
    justifyContent: "center",
  },
  pageTitle: {
    alignSelf: "flex-start",
    marginTop: "5%",
    marginLeft: "5%",
  },
  form: {
    flex: 1,
    width: "90%",
    alignItems: "center",
    flexDirection: "column",
  },
  question: {
    alignSelf: "flex-start",
    marginTop: "10%",
    marginBottom: "2%",
    color: "#474747",
    fontSize: 18,
    fontWeight: "bold",
  },
  healthInfoContainer: {
    marginTop: "10%",
    padding: "5%",
    alignItems: "center",
    justifyContent: "center",
    backgroundColor: "#eafbff",
    borderRadius: 24,
  },
  informationalTextHeader: {
    fontSize: 18,
    color: "#474747",
    textAlign: "center",
    fontWeight: "bold",
  },
  informationalText: {
    fontSize: 16,
    color: "#474747",
    textAlign: "left",
    marginTop: "5%",
    lineHeight: 26,
  },
  learnMoreText: {
    fontSize: 16,
    color: "#0079d3",
    textDecorationLine: "underline",
  },
  saveButton: {
    alignSelf: "flex-end",
    marginBottom: "10%",
    minWidth: "90%",
  },
});

export default DietGoalsScreen;
