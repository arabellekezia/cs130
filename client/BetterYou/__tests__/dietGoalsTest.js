import {render, cleanup, fireEvent} from '@testing-library/react-native';
import React from 'react';
import DietGoalsScreen from "../app/Screens/DietGoalsScreen"

describe('Diet goals input form', () => {
  let wrapper;
  beforeEach(() => {
      wrapper = render(<DietGoalsScreen />);
    });
  afterEach(() => {
      cleanup();
      wrapper = null;
    });
  it('renders correctly', () => {
      expect(wrapper).toMatchSnapshot();
  });

  it("No error message is displayed with valid input.", () => {
    const {getByText, queryByTestId, getByPlaceholderText} = wrapper;

    const goalTextInput = getByPlaceholderText("2000 cal"); 
    const enteredGoal = "1800";
    fireEvent(goalTextInput, 'onChangeText', enteredGoal);
    
    const saveButton = getByText("Save");
    fireEvent.press(saveButton);
  
    const validationError = queryByTestId("error-message");
    expect(validationError).toBe(null);
  });

  it("Shows error message if daily calorie budget is less than 1000 calories (not reasonable input)", () => {
    const {getByText, queryByTestId, getByPlaceholderText} = wrapper;

    const goalTextInput = getByPlaceholderText("2000 cal"); 
    const enteredGoal = "800";
    fireEvent(goalTextInput, 'onChangeText', enteredGoal);
    
    const saveButton = getByText("Save");
    fireEvent.press(saveButton);
  
    const validationError = queryByTestId("error-message");
    expect(validationError).toBeTruthy();
    
    expect(validationError.props.children)
        .toBe("Your daily calorie budget must be greater than 1000 calories.");
  });

  it("Shows error message if the input is not a number.", () => {
    const {getByText, queryByTestId, getByPlaceholderText} = wrapper;

    const goalTextInput = getByPlaceholderText("2000 cal"); 
    const enteredGoal = "2000,";
    fireEvent(goalTextInput, 'onChangeText', enteredGoal);
    
    const saveButton = getByText("Save");
    fireEvent.press(saveButton);
  
    const validationError = queryByTestId("error-message");
    expect(validationError).toBeTruthy();
    
    expect(validationError.props.children)
        .toBe("Your daily calorie budget must be a number.");
  });
});

