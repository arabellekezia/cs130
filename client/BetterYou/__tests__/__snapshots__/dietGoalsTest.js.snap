// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Diet goals input form renders correctly 1`] = `
<RCTScrollView
  collapsable={false}
  contentContainerStyle={
    Object {
      "flex": 1,
    }
  }
  keyboardShouldPersistTaps="handled"
  onGestureHandlerEvent={[Function]}
  onGestureHandlerStateChange={[Function]}
>
  <View>
    <RCTSafeAreaView
      emulateUnlessSupported={true}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "white",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontSize": 30,
              "fontWeight": "bold",
            },
            Object {
              "alignSelf": "flex-start",
              "marginLeft": "5%",
              "marginTop": "5%",
            },
          ]
        }
      >
        Diet Goals
      </Text>
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "column",
            "width": "90%",
          }
        }
      >
        <Text
          style={
            Array [
              Object {},
              Object {
                "alignSelf": "flex-start",
                "color": "#474747",
                "fontSize": 18,
                "fontWeight": "bold",
                "marginBottom": "2%",
                "marginTop": "10%",
              },
            ]
          }
        >
          What is your daily calorie budget?
        </Text>
        <View
          style={
            Object {
              "backgroundColor": "#efefef",
              "borderRadius": 10,
              "flexDirection": "row",
              "marginVertical": 10,
              "padding": 15,
              "width": "100%",
            }
          }
        >
          <TextInput
            allowFontScaling={true}
            keyboardType="numeric"
            maxLength={25}
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            placeholder="2000 cal"
            rejectResponderTermination={true}
            style={
              Object {
                "fontSize": 18,
              }
            }
            underlineColorAndroid="transparent"
          />
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#eafbff",
              "borderRadius": 24,
              "justifyContent": "center",
              "marginTop": "10%",
              "padding": "5%",
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#474747",
                "fontSize": 18,
                "fontWeight": "bold",
                "textAlign": "center",
              }
            }
          >
            New to calorie counting?
          </Text>
          <Text
            style={
              Object {
                "color": "#474747",
                "fontSize": 16,
                "lineHeight": 26,
                "marginTop": "5%",
                "textAlign": "left",
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#474747",
                  "fontSize": 16,
                  "lineHeight": 26,
                  "marginTop": "5%",
                  "textAlign": "left",
                }
              }
            >
              Use the following 
            </Text>
            <Text
              onPress={[Function]}
              style={
                Object {
                  "color": "#0079d3",
                  "fontSize": 16,
                  "textDecorationLine": "underline",
                }
              }
            >
              calorie calculator
            </Text>
            <Text>
               as a guideline to determine your daily calorie intake based on your weight loss goals!
            </Text>
          </Text>
        </View>
      </View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "alignSelf": "flex-end",
              "backgroundColor": "#007AFF",
              "borderRadius": 10,
              "justifyContent": "center",
              "marginBottom": "10%",
              "minWidth": "90%",
              "paddingHorizontal": 30,
              "paddingVertical": 12,
            }
          }
        >
          <Text
            style={
              Array [
                Object {},
                Object {
                  "color": "white",
                  "fontSize": 18,
                  "fontWeight": "600",
                },
              ]
            }
          >
            Save
          </Text>
        </View>
      </View>
    </RCTSafeAreaView>
  </View>
</RCTScrollView>
`;
